crates/sui-framework/packages/sui-framework/sources/derived_object.move

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Enables the creation of objects with deterministic addresses derived from a parent object's UID.
+/// This module provides a way to generate objects with predictable addresses based on a parent UID
+/// and a key, creating a namespace that ensures uniqueness for each parent-key combination,
+/// which is usually how registries are built.
+///
+/// Key features:
+/// - Deterministic address generation based on parent object UID and key
+/// - Derived objects can exist and operate independently of their parent
+///
+/// The derived UIDs, once created, are independent and do not require sequencing on the parent
+/// object. They can be used without affecting the parent. The parent only maintains a record of
+/// which derived addresses have been claimed to prevent duplicates.
+module sui::derived_object;
+
+use sui::dynamic_field as df;
+
+/// Tries to create an object twice with the same parent-key combination.
+const EObjectAlreadyExists: u64 = 0;
+/// Tries to restore an object that does not exist for the supplied parent.
+const EInvalidParent: u64 = 1;
+/// Tries to use functionality that is not supported yet.
+const ENotSupported: u64 = 2;
+
+/// Added as a DF to the parent's UID, to mark an ID as claimed.
+public struct Claimed(ID) has copy, drop, store;
+
+/// An internal key to protect from generating the same UID twice (e.g. collide with DFs)
+public struct DerivedObjectKey<K: copy + drop + store>(K) has copy, drop, store;
+
+/// Claim a deterministic UID, using the parent's UID & any key.
+public fun claim<K: copy + drop + store>(parent: &mut UID, key: K): UID {
+    let addr = derive_address(parent.to_inner(), key);
+    let id = addr.to_id();
+
+    // If the UID has never been claimed, we can generate it and return early.
+    if (!df::exists_(parent, Claimed(id))) {
+        let uid = object::new_uid_from_hash(addr);
+
+        // We save the value as `Option<UID>` to allow us to have "restore" functionality for
+        // a derived UID.
+        df::add<_, Option<UID>>(parent, Claimed(id), option::none());
+
+        return uid
+    };
+
+    // IF the UID has been restored, we can re-use it.
+    let existing_uid = df::borrow_mut<_, Option<UID>>(parent, Claimed(id));
+
+    assert!(existing_uid.is_some(), EObjectAlreadyExists);
+
+    abort ENotSupported
```

Comment:
Interestingly enough, even without id-leak verifier change, one could use the full flow `claim -> return -> restore` to receive TTO'd objects. So if I own the "registry UID", I can do `claim -> return` without using this `UID` to create new objects.